beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+// attempt to hold a file lock to guarantee exclusive control of the data column storage directory, so this should only be
+// initialized once per beacon node.
+func NewDataColumnStorage(ctx context.Context, opts ...DataColumnStorageOption) (*DataColumnStorage, error) {
+	storage := &DataColumnStorage{
+		DataColumnFeed: new(event.Feed),
+		muChans:        make(map[[fieldparams.RootLength]byte]*muChan),
+	}
+
+	for _, o := range opts {
+		if err := o(storage); err != nil {
+			return nil, errors.Wrap(err, "failed to create data column storage")
+		}
+	}
+
+	// Allow tests to set up a different fs using WithFs.
+	if storage.fs == nil {
+		if storage.base == "" {
+			return nil, errNoBasePath
+		}
+
+		storage.base = path.Clean(storage.base)
+		if err := file.MkdirAll(storage.base); err != nil {
+			return nil, errors.Wrapf(err, "failed to create data column storage at %s", storage.base)
+		}
+
+		storage.fs = afero.NewBasePathFs(afero.NewOsFs(), storage.base)
+	}
+
+	storage.cache = newDataColumnStorageSummaryCache()
+
+	async.RunEvery(ctx, prunePeriod, func() {
+		storage.pruneMu.Lock()
+		defer storage.pruneMu.Unlock()
+
+		storage.prune()
+	})
+
+	return storage, nil
+}
+
+// WarmCache warms the cache of the data column filesystem.
+// It holds the database (read) lock for all the time it is running.
+func (dcs *DataColumnStorage) WarmCache() {
+	start := time.Now()
+	log.Info("Data column filesystem cache warm-up started")
+
+	dcs.pruneMu.Lock()
+	defer dcs.pruneMu.Unlock()
+
+	highestStoredEpoch := primitives.Epoch(0)
+
+	// Walk the data column filesystem to warm up the cache.
+	if err := afero.Walk(dcs.fs, ".", func(path string, info os.FileInfo, fileErr error) (err error) {
+		if fileErr != nil {
+			return fileErr
+		}
+
+		// If not a leaf, skip.
+		if info.IsDir() {
+			return nil
+		}
+
+		// Extract metadata from the file path.
+		fileMetadata, err := extractFileMetadata(path)
+		if err != nil {
+			log.WithError(err).Error("Error encountered while extracting file metadata")
+		}
+
+		// Open the data column filesystem file.
+		file, err := dcs.fs.Open(path)
+		if err != nil {
+			log.WithError(err).Error("Error encountered while opening data column filesystem file")
+			return nil
+		}
+
+		// Close the file.
+		defer func() {
+			// Overwrite the existing error only if it is nil, since the close error is less important.
+			closeErr := file.Close()
+			if closeErr != nil && err == nil {
+				err = closeErr
+			}
+		}()
+
+		// Read the metadata of the file.
+		metadata, err := dcs.metadata(file)
+		if err != nil {
+			log.WithError(err).Error("Error encountered while reading metadata from data column filesystem file")
+			return nil
+		}
+
+		// Check the indices.
+		indices := make([]uint64, 0, len(metadata.indices))
+		for index, i := range metadata.indices {
+			// Skip if the data column is not saved.
+			if i < limit {
```

Comment:
I get it, if the high bit is set, the value will be > 128. But the name `limit` for this variable threw me off, it's a little confusing to read. Maybe something like `nonZeroOffset` would be clear?

Reply:
Yes I struggled to find a good name. I was not happy with `limit` as well.

Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/584ef60ab0963bf85db19db637e131610ef9f4e6.